### PR TITLE
PRO-1172 manager tables, manager context menus, etc

### PR DESCRIPTION
**Pull request by stuartromanek**
#### Main ticket
- https://linear.app/apostrophecms/issue/PRO-1172/general-manager-sprucing

#### Related tickets
- https://linear.app/apostrophecms/issue/PRO-1156/signal-a-document-is-draft-only

#### Main points
- Adds context menus for documents in piece managers and page manager
- Wires up context actions for docs (archive, unarchive, edit, preview, discard draft)
- Adds generic AposLabel component
- Adds labels to docs in managers indicating unpublished, active draft, archived

### CODE REVIEW

File: modules/@apostrophecms/admin-bar/ui/apos/components/TheAposAdminBarMenu.vue [hunk 1]
```diff
@@ -3,7 +3,7 @@
     <li class="apos-admin-bar__item" v-if="pageTree">
       <AposButton
         type="subtle" label="Page Tree"
-        icon="file-tree-icon" class="apos-admin-bar__btn"
```
Comment (stuartromanek) [inline modules/@apostrophecms/admin-bar/ui/apos/components/TheAposAdminBarMenu.vue:6]: unrelated

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue [hunk 2]
```diff
@@ -106,6 +108,7 @@ import AposModalModifiedMixin from 'Modules/@apostrophecms/modal/mixins/AposModa
 import AposModalTabsMixin from 'Modules/@apostrophecms/modal/mixins/AposModalTabsMixin';
 import AposEditorMixin from 'Modules/@apostrophecms/modal/mixins/AposEditorMixin';
 import AposPublishMixin from 'Modules/@apostrophecms/ui/mixins/AposPublishMixin';
+import AposArchiveMixin from 'Modules/@apostrophecms/ui/mixins/AposArchiveMixin';
```
Comment (stuartromanek) [inline modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue:111]: generic UI methods for archiving and unarchiving documents

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue [hunk 3]
```diff
@@ -23,11 +23,13 @@
...
         :can-archive="canArchive"
         :can-copy="!!docId"
+        :can-preview="canPreview"
         :is-published="!!published"
         :can-save-draft="true"
         @saveDraft="saveDraft"
+        @preview="preview"
         @discardDraft="onDiscardDraft"
-        @moveToArchive="onMoveToArchive"
+        @archive="onArchive"
```
Comment (stuartromanek) [inline modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue:32]: rename

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue [hunk 4]
```diff
@@ -605,35 +622,10 @@ export default {
         return this.$refs[field.group.name][0];
       }
     },
-    async onMoveToArchive(e) {
```
Comment (stuartromanek) [inline modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue:608]: moved to mixin

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocMoreMenu.vue [hunk 5]
```diff
@@ -5,6 +5,8 @@
     :disabled="disabled"
     menu-placement="bottom-end"
     @item-clicked="menuHandler"
+    @open="$emit('menuOpen')"
+    @close="$emit('menuClose')"
```
Comment (stuartromanek) [inline modules/@apostrophecms/doc-type/ui/apos/components/AposDocMoreMenu.vue:9]: let our UI know if the menu is left open so we can do things like keep it visible

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 6]
```diff
@@ -18,7 +18,7 @@
       </transition>
       <transition :name="transitionType" @after-leave="$emit('inactive')">
         <div
-          v-if="modal.showModal"
+          v-if="modal.showModal" :class="innerClasses"
```
Comment (stuartromanek) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:21]: unrelated, provide dimensions for slide in modals

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 7]
```diff
@@ -122,6 +122,13 @@ export default {
       }
       return classes.join(' ');
     },
+    innerClasses() {
+      const classes = [];
+      if (this.modal.width) {
+        classes.push(`apos-modal__inner--${this.modal.width}`);
+      };
+      return classes;
+    },
```
Comment (stuartromanek) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:131]: unrelated, provide dimensions for slide in modals

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 8]
```diff
@@ -236,7 +243,7 @@ export default {
     height: calc(100vh - #{$spacing-double * 2});
     border-radius: var(--a-border-radius);
     background-color: var(--a-background-primary);
-    border: 1px solid var(--a-base-4);
+    border: 1px solid var(--a-base-9);
```
Comment (stuartromanek) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:246]: unrelated sprucing

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 9]
```diff
@@ -255,6 +262,18 @@ export default {
...
+        max-width: 66%;
+      }
+    }
+
+    &.apos-modal__inner--half {
+      @media screen and (min-width: 800px) {
+        max-width: 50%;
+      }
+    }
+
```
Comment (stuartromanek) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:276]: unrelated slide in modal sizes

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 10]
```diff
@@ -333,7 +352,7 @@ export default {
   }
 
   .apos-modal__header__main {
-    border-bottom: 1px solid var(--a-base-4);
+    border-bottom: 1px solid var(--a-base-9);
```
Comment (stuartromanek) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:355]: unrelated sprucing

File: modules/@apostrophecms/modal/ui/apos/mixins/AposDocsManagerMixin.js [hunk 11]
```diff
@@ -99,6 +99,12 @@ export default {
     }
   },
   methods: {
+    preview(doc) {
+      window.open(doc._url, '_blank').focus();
+    },
+    findDoc(id, docs) {
```
Comment (stuartromanek) [inline modules/@apostrophecms/modal/ui/apos/mixins/AposDocsManagerMixin.js:105]: utility

File: modules/@apostrophecms/notification/ui/apos/components/AposNotification.vue [hunk 12]
```diff
@@ -194,10 +194,11 @@ export default {
 
     & /deep/ button {
       @include apos-button-reset();
+      padding: 0 3px;
+      font-size: inherit;
```
Comment (stuartromanek) [inline modules/@apostrophecms/notification/ui/apos/components/AposNotification.vue:198]: smooth edges of notification style updates

File: modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManager.vue [hunk 13]
```diff
@@ -18,7 +18,6 @@
       />
     </template>
     <template #primaryControls>
-      <!-- TODO Make sure this gets positioned correctly after Vue3/teleport -->
```
Comment (stuartromanek) [inline modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManager.vue:21]: fixed, but without vue3 or teleport

File: modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManager.vue [hunk 14]
```diff
@@ -160,29 +171,6 @@ export default {
...
-          data[column.name] = piece[column.name];
-        });
-
-        data._id = piece._id;
-
-        items.push(data);
-      });
-
-      return items;
-    },
```
Comment (stuartromanek) [inline modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManager.vue:185]: This didn't seem like it providing much. It was getting in the way of checking non-header doc properties, we never modified it, i just subbed in the `pieces` prop where it was used

File: modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManagerDisplay.vue [hunk 15]
```diff
@@ -23,12 +23,21 @@
...
+            More Operations
+          </component>
+        </th>
       </tr>
       <tr
         class="apos-table__row"
         v-for="item in items"
         :key="item._id"
         :class="{'is-selected': false }"
+        @mouseover="over(item._id)" @mouseout="out(item._id)"
```
Comment (stuartromanek) [inline modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManagerDisplay.vue:40]: provide interaction details on the whole row

File: modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManagerDisplay.vue [hunk 16]
```diff
@@ -104,8 +124,26 @@ export default {
...
-    'updated'
+    'updated',
+    'preview',
+    'copy',
+    'discardDraft',
+    'archive',
+    'unarchive'
   ],
+  data() {
+    const state = {
```
Comment (stuartromanek) [inline modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManagerDisplay.vue:135]: keep track of hover states of rows to pass into child components

File: modules/@apostrophecms/ui/ui/apos/components/AposContextMenu.vue [hunk 17]
```diff
@@ -16,7 +16,7 @@
       <!-- TODO refactor buttons to take a single config obj -->
       <AposButton
         class="apos-context-menu__btn"
-        @click="buttonClicked($event)"
+        @click.stop="buttonClicked($event)"
```
Comment (stuartromanek) [inline modules/@apostrophecms/ui/ui/apos/components/AposContextMenu.vue:19]: stop click propagation on context menus to avoid confusing their parents

File: modules/@apostrophecms/ui/ui/apos/components/AposTreeRows.vue [hunk 18]
```diff
@@ -287,15 +300,12 @@ export default {
         }
       ];
 
-      if (this.options.ghostUnpublished) {
```
Comment (stuartromanek) [inline modules/@apostrophecms/ui/ui/apos/components/AposTreeRows.vue:290]: indicate unpublished w a label

File: modules/@apostrophecms/modal/ui/apos/mixins/AposDocsManagerMixin.js (hunk 11 above)
Comment (boutell) [inline modules/@apostrophecms/modal/ui/apos/mixins/AposDocsManagerMixin.js:105]: I would go with findDocById(docs, id). We generally put the haystack before the needle.

File: modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManager.vue [hunk 19]
```diff
@@ -276,6 +264,35 @@ export default {
...
+    onPreview(id) {
+      this.preview(this.findDoc(id, this.pieces));
+    },
+    async onArchive(id) {
+      const piece = this.findDoc(id, this.pieces);
+      if (await this.archive(this.options.action, id, !!piece.lastPublishedAt)) {
+        apos.bus.$emit('content-changed');
+      }
+    },
+    async onUnarchive(id) {
```
Comment (boutell) [inline modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManager.vue:276]: Restore

File: modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManager.vue [hunk 20]
```diff
@@ -276,6 +264,35 @@ export default {
...
+      const piece = this.findDoc(id, this.pieces);
+      if (await this.unarchive(this.options.action, id, !!piece.lastPublishedAt)) {
```
Comment (boutell) [inline modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManager.vue:278]: Restore

File: modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManagerDisplay.vue (hunk 16 above)
Comment (boutell) [reply modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManagerDisplay.vue:135]: restore

**Review by boutell — CHANGES_REQUESTED** [review]
This looks great, my only concern is that the term "restore" is already in use and is nicer than "unarchive." We use it on the front end and it's a good verb, so I reckon we should on the back end too.

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue (hunk 2 above)
Comment (localghost443) [reply modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue:111]: : _)

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue [hunk 21]
```diff
@@ -275,6 +279,13 @@ export default {
     canPreviewDraft() {
       return !this.docId && this.moduleOptions.previewDraft;
     },
+    canPreview() {
+      if (this.original) {
+        return !!this.original._url;
+      } else {
+        return null;
```
Comment (abea) [inline modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue:286]: `false` seems more appropriate, fwiw

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue [hunk 22]
```diff
@@ -409,6 +420,12 @@ export default {
     }
   },
   methods: {
+    async preview() {
+      if (!await this.confirmAndCancel()) {
+        return;
+      }
+      window.location = this.original._url;
```
Comment (abea) [inline modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue:427]: Here we're opening in the same tab. In the manager more menu we open in a new tab. Is that intentional? It seems it should be consistent.

File: modules/@apostrophecms/piece-type/index.js [hunk 23]
```diff
@@ -29,20 +29,18 @@ module.exports = {
...
-        visibility: {
-          label: 'Visibility'
+        labels: {
+          name: 'labels',
+          label: '',
+          component: 'AposCellLabels'
         },
-        // Automatically hidden if none of the pieces
-        // actually have a URL
-        _url: {
```
Comment (abea) [inline modules/@apostrophecms/piece-type/index.js:43]: wat

File: modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManager.vue (hunk 13 above)
Comment (abea) [reply modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManager.vue:21]: magic.

File: modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManager.vue (hunk 14 above)
Comment (abea) [reply modules/@apostrophecms/piece-type/ui/apos/components/AposPiecesManager.vue:185]: 🆒 I think this is from when the table markup was in this component and the property was called `rows`.

File: modules/@apostrophecms/piece-type/index.js (hunk 23 above)
Comment (abea) [reply modules/@apostrophecms/piece-type/index.js:43]: 🆗

**Review by abea — CHANGES_REQUESTED** [review]
The Archive "page" itself is getting the "Archived" label. If we're planning to move it to its own tab that's fine. If that's not happening before stable then it seems weird.

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue (hunk 22 above)
Comment (boutell) [reply modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue:427]: I don't think you should get hauled off to a new tab if the thing you're navigating to is the thing you just created/edited.

Whereas I can at least see that in the case of the manage view.

**Review by boutell — DISMISSED** [review]
Approving re: the concerns I raised myself.

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue (hunk 22 above)
Comment (stuartromanek) [reply modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue:427]: i had Reasons but they seem not great now .. gut says everyone opens a new tab? I would also axe the confirm-discarding-changes in that change. Make sense?

Comment (stuartromanek) [reply modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue]: opens a new tab now

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue (hunk 21 above)
Comment (stuartromanek) [reply modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue]: `true`

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue (hunk 22 above)
Comment (stuartromanek) [reply modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue]: @boutell the saveDraftAndPreview button still operates as it did (same tab).. is that what you mean?

Comment (stuartromanek) [reply modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue:427]: @abea @boutell I'm backed this out to the original way I had it, I think it does make sense:
- In a manager your browsing lots of things, previewing one does not indicate your done with all of them
- In the editor you're focused on one document, and potentially you've made edits to the fields inside. Opening a new tab splits your attention between a thing your editing and a preview that can't reflect those changes. I think in that case it makes sense to force the editor to make a decision about anything they've done up to that point and then change their entire scene to the in-context preview

**Review by boutell — DISMISSED** [review]

**Review by abea — APPROVED** [review]

**Comment by abea** [comment]
😬 Tom renamed some files at my suggestion (not only about pieces anymore). The date one is not on me, but is correct.
